user-operator/src/main/java/io/strimzi/operator/user/operator/KafkaUserOperator.java

```diff
@@ -146,7 +156,117 @@ protected void createOrUpdate(Reconciliation reconciliation, KafkaUser kafkaUser
                 secretOperations.reconcile(namespace, user.getSecretName(), desired),
                 aclOperations.reconcile(KafkaUserModel.getTlsUserName(userName), tlsAcls),
                 aclOperations.reconcile(KafkaUserModel.getScramUserName(userName), scramAcls))
-                .map((Void) null).setHandler(handler);
+                .compose(reconciliationResult -> {
+                    Future<Void> fut = Future.future();
+                    Condition readyCondition;
+                    KafkaUserStatus userStatus = new KafkaUserStatus();
+
+                    if (kafkaUser.getMetadata().getGeneration() != null)    {
+                        userStatus.setObservedGeneration(kafkaUser.getMetadata().getGeneration());
+                    }
+
+                    if (reconciliationResult.succeeded()) {
+                        readyCondition = new ConditionBuilder()
+                                .withNewLastTransitionTime(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ").format(dateSupplier()))
+                                .withNewType("Ready")
+                                .withNewStatus("True")
+                                .build();
+                    } else {
+                        readyCondition = new ConditionBuilder()
+                                .withNewLastTransitionTime(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ").format(dateSupplier()))
+                                .withNewType("NotReady")
+                                .withNewStatus("True")
+                                .withNewReason(fut.cause().getClass().getSimpleName())
+                                .withNewMessage(fut.cause().getMessage())
+                                .build();
+                    }
+
+                    Credential credential = null;
+                    if (desired != null) {
+                        credential = new CredentialBuilder()
+                                .withNewCredentials("")
+                                .withCredentialsSecret(desired.getMetadata().getName())
+                                .build();
+                    }
```

Comment:
I would say yes, it does. @tombentley can you confirm, please?

Reply:
C'mon guys, you can figure this out for yourselves...

The documentation of `CompositeFuture.join()` reads:

> Return a composite future, succeeded when all futures are succeeded, **failed when any future is failed**.
> 
> It always wait until all its futures are completed and will not fail as soon as one of f1 or f2 fails.

And the JavaDoc for `Future.compose(Function)` reads:

> Compose this future with a mapper function.
When this future (the one on which compose is called) succeeds, the mapper will be called with the completed value and this mapper returns another future object. This returned future completion will complete the future returned by this method call.
If the mapper throws an exception, the returned future will be failed with this exception.
**When this future fails, the failure will be propagated to the returned future and the mapper will not be called.**

Emphasis mine.

If you want to run code when a future fails you use:

* `Future.setHandler()`, which can give you something like a `finally` clause
* `Future.recover()`, which ends up being a bit similar to a `catch(Throwable) {   }`

`Future.otherwise()` is another option.

Reply:
@tombentley @stanlyDoge Just double checking ;) because I don't think the credential field should be updated upon the failure unless it is to be set to null. I think the way the code is now, if the desired secret fails to be created, the credential status field still gets populated with that secret.

Reply:
The way I did it for the KafkaStatus is that I actually assemble the status during the reconciliation and not just at the end. So the values are set if we created the underlying resources. Maybe the same can be done here - to set the secret name when the secret is created?